server/src/grpc/remote_engine_service/mod.rs

```diff
@@ -97,6 +155,139 @@ impl<Q: QueryExecutor + 'static> RemoteEngineServiceImpl<Q> {
         Ok(ReceiverStream::new(rx))
     }
 
+    async fn deduped_stream_read_internal(
+        &self,
+        request: Request<ReadRequest>,
+    ) -> Result<ReceiverStream<Result<RecordBatch>>> {
+        let instant = Instant::now();
+        let ctx = self.handler_ctx();
+        let (tx, rx) = mpsc::channel(STREAM_QUERY_CHANNEL_LEN);
+
+        let request = request.into_inner();
+        let table_engine::remote::model::ReadRequest {
+            table,
+            read_request,
+        } = request.clone().try_into().box_err().context(ErrWithCause {
+            code: StatusCode::BadRequest,
+            msg: "fail to convert read request",
+        })?;
+
+        let request_key = RequestKey::new(
+            table.table,
+            read_request.predicate.clone(),
+            read_request.projected_schema.projection(),
+            read_request.order,
+        );
+
+        let mut guard = match self
+            .request_notifiers
+            .as_ref()
+            .unwrap()
```

Comment:
Maybe we can declare the `request_notifiers` as the params of `deduped_stream_read_internal` to avoid call `as_ref().unwrap()` again and again.